### ISSUE TITLE
UnicodeEncodeError causes deencapsulation to fail when surrogates are used. 

### ISSUE THREAD

**Issue by TheElementalOfDestruction** [seamustuohy/RTFDE#6]
Traceback:
-----------

```
Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\site-packages\RTFDE\deencapsulate.py", line 134, in deencapsulate
    self.content = self._deencapsulate_from_tree()
  File "C:\Program Files\Python39\lib\site-packages\RTFDE\deencapsulate.py", line 180, in _deencapsulate_from_tree
    decoded_tree = RTFUnicodeDecoder().visit_topdown(self.doc_tree)
  File "C:\Program Files\Python39\lib\site-packages\RTFDE\transformers.py", line 59, in visit_topdown
    char = chr(bytearray(chr(65536+char_num).encode())[-self.cur_uc[-1]])
UnicodeEncodeError: 'utf-8' codec can't encode character '\ud83d' in position 0: surrogates not allowed
```

Description:
-------------
Unicode has special characters known as surrogates, characters that expect another full character to come after them to determine their full value. This was designed for compatibility with UTF-16 which only allowed for characters up to a value of 0xFFFF. These should be decodable if you put them together and then decode using utf-16 in a special way. 

I found that emails that contain emojis but no HTML body will try to deencapsulate the html and fail with this error. For context, this is the section it tried to parse when it failed.
```rtf
{\*\htmltag148 <span style="font-family:&quot;Segoe UI Emoji&quot;,sans-serif">}\htmlrtf {\f4 \htmlrtf0 \u-10179 ?\u-8694 ?
{\*\htmltag156 </span>}\htmlrtf }\htmlrtf0  
```

You can see this answer on stack overflow for details on how to possibly handle it. I recommend checking the value before encoding (unfortunate as that is) or simply catching the encoding error and then doing this step to compensate for it. 

https://stackoverflow.com/a/54549164/7723054

PS: I'm currently trying to see if I can actually generate a full MSG file for you to use to analyze this issue. I'll leave a response if I can make one.

**Comment by TheElementalOfDestruction** [comment]
I managed to generate an msg file for you, stripped of personal data of course. Funnily, this is the character that is causing all of the commotion: 😊 
[Demonstration of surrogates.zip](https://github.com/seamustuohy/RTFDE/files/8883154/Demonstration.of.surrogates.zip)

**Comment by seamustuohy** [comment]
Oh no! Surrogate pairs... UTF-16 was such a mistake. This looks like it will require some post-processing of the grammar tree. I'm  updating the grammar to be more nuanced and easier to edit on a granular level in response to another issue. I think that by adding identification of Unicode encoding to the grammar it will make post processing possible without slowing the parsing too much.

**Comment by seamustuohy** [comment]
I want to congratulate you on this issue. It not only exposed a the aforementioned surrogate pairs issue. It also may be the key to the whitespace concerns. Surrogate pairs were apparently the issue that caused me to give up on the previous whitespace issues. Once I get these under control (I'm currently contacting some MS folks about this to try and figure out the right way to handle this undocumented "feature") I think it will address the whitespace concerns as well.

**Comment by TheElementalOfDestruction** [comment]
Oh that sounds like fun.

If by "undocumented feature" you mean surrogate pairs, I think that's just supposed to be handled (just a wild guess) by having the read data be considered utf-16 when decoding them.

I actually managed to do the opposite and create surrogate pairs of Unicode rtf tags by encoding into utf-16-le, reading the 4 bytes using struct as 2 signed shorts (I actually decided not to count on the format only using 2 characters myself and allowed for more) and then putting those numbers for the Unicode character tags. It rendered correctly in both wordpad and word, so the opposite approach of reading a series of Unicode character things, packing all the ones in a row into bytes, then decoding as utf-16-le should do the trick.

**Comment by seamustuohy** [comment]
Haha, The undocumented feature isn't the surrogate pairs. It's the way that RTF encapsulation handles escaping around surrogate pairs. The current modifications (not yet pushed) can handle the surrogate pairs. The problem is handling the question marks used to escape and the spaces within/without correctly to produce an identical RTF input from the output.

**Comment by TheElementalOfDestruction** [comment]
This is the relevent section I found in the docs about unicode characters. Basically just if you understand the unicode character, skip the next character (which may or may not have a single space between the last digit and that character, and may not be a ? or even a single character in the RTF file, which is nuts). So if you don't actually care about making the RTF invalid you could just purge those characters before deencapsulating?

![image](https://user-images.githubusercontent.com/24580325/176328271-11266941-6bc0-452c-aa32-370acd95af46.png)

**Comment by TheElementalOfDestruction** [comment]
My guess is that it really doesn't do much other than acknowladge that it's a surrogate value (there is a list of them online) and then just store it in a buffer or something. If it finds anything that wouldn't be the other part of the pair, it would probably just return the two values as directly as possible or throw an error. Actually, why guess when you can test what an official parser would do about invalid pairings (say, if you put a random character between them, or no second character). Now I'm interested and think I'm going to find out

**Comment by seamustuohy** [comment]
Yes, I understand that. I've read it, and multiple other versions of the standard searching for the answer. And, the code accounts for that. But, that's not the entire picture. There is an undocumented component where a paired escape character surrounds a set of characters. Before and after. It's what creates the extra spaces. I need to identify the specific mechanism so I don't add additional spaces into the deencapsulation.

**Comment by TheElementalOfDestruction** [comment]
Well if you let me know what the component is, I might be able to do my own research, if you like.

**Comment by seamustuohy** [comment]
Take a look at \u-8694 in the code you submitted for an example.

**Comment by seamustuohy** [comment]
It's just an example of a single character. But, I want to get all the edge cases so I don't add more instability by addressing only what I can see in that example.

**Comment by TheElementalOfDestruction** [comment]
Not sure I see what about it is undocumented. I seem to be understanding exactly what it is saying. Oh well

**Comment by seamustuohy** [comment]
Ok, that gives me hope that I've just been misunderstanding the spec when I
read it. I'll take a look with fresh eyes tomorrow.

**Comment by seamustuohy** [comment]
Yup, you were correct. Thanks for pushing back. At some point in the current parser I had it merging the tokens incorrectly so that the question marks were being joined with the Unicode character that followed them (instead of the one that they followed.) So, the structure was jumbled. That fix in the pipeline should make this /somewhat/ easy to address.

**Comment by seamustuohy** [comment]
It looks like this issue can be addressed by adding [surrogatepass error handling](https://docs.python.org/3/library/codecs.html#error-handlers) to the [section of code that handles unicode character encoding](https://github.com/seamustuohy/RTFDE/blob/main/RTFDE/transformers.py#L59). 

The basic unittest I've written to handle the surrogate-pair that you attached above decodes that character as the "½" character. Let me know if that is correct. If so, then I'll roll this fix into the other set of fixes I'm doing right now. 

This line of code will look like the below when it is done.
```python
char = chr(bytearray(chr(65536+char_num).encode(errors='surrogatepass'))[-self.cur_uc[-1]])
```

**Comment by TheElementalOfDestruction** [comment]
Unfortunately, that is not the right character. The character is meant to be a smiling face. Here is an example of some code that actually unpacked it properly. I manually copied the character values in, so of course you would have to acquire them through the code, but this works to give you the proper codepoint:
```py
>>> import struct
>>> # The text was "\u-10179 ?\u-8694 ?" to be decoded.
>>> print(hex(ord(struct.pack('<hh', -10179, -8694).decode('utf-16-le'))))
0x1f60a
```
This codepoint matches in that the character represented is `\U0001F60A`.

**Comment by TheElementalOfDestruction** [comment]
Ah, I think I see the problem. So this section:
```py
chr(65536+-10179).encode(errors='surrogatepass')
```
Actually gives this bytes value: `b'\xed\xa0\xbd'`, and wouldn't you know it, the 1/2 symbol is U+00BD, the last byte in that. So it looks like your code grabbed the last bit and then ran with it instead of combining with the next character. Perhaps I'm misunderstanding what the rest of that code doing, but that's what it looks like to be.

**Comment by seamustuohy** [comment]
Did you manually copy the character values into an existing RTF document or, did you encapsulate this using Microsoft software? I just want to make sure I understand the test file and if I'm handling \uc chars wrong, or if I might also be seeing an artifact in my code related to the unicode byte characters.

**Comment by TheElementalOfDestruction** [comment]
The example file I made by opening outlook at pasting the character in, saving that. Opening in outlook also shows the body correctly.

However, I also manually copied the characters and got an identical result. After testing. I understand the problem: the `\ucN` tag is not being handled right. The language in the documentation is weird, and the reasoning complex, but it seems the best way to treat it is to just ignore N numbers of characters after the Unicode tag. Testing the exact same single Unicode character tag with 1 or 2 question marks after it outputted a single question mark after the Unicode character for `\uc1` and no question makes after the character for `\uc2`. I don't know what your code was trying to do by indexing a byte array, but that doesn't seem to be the proper behavior.

**Comment by TheElementalOfDestruction** [comment]
Here is a file that can be opened in wordpad or word to help demonstrate the effect of the `\ucN` tag on unicode characters. Each visual line of the file is on it's own line in the .rtf file source as well for easy reading and comprehension.

[Document.zip](https://github.com/seamustuohy/RTFDE/files/9026282/Document.zip)

**Comment by seamustuohy** [comment]
I just want to say thank you so much for these test files. I have had to build this library entirely without access to a windows system. So, I've not been able to create the test files I needed to verify that I was interpreting the spec correctly. It's surprisingly hard to find encapsulated rtf files with various features contained within them. Having real test-cases to use has been a blessing.

**Comment by TheElementalOfDestruction** [comment]
No problem. Unfortunately for this kind of testing it's rather necessary.

**Comment by TheElementalOfDestruction** [comment]
By the way, I have access to a substantial amount of, unfortunately, confidential MSG files where most have encapsulated HTML in the RTF stream, so I have the ability to rather easily test against a massive battery of files and a lot of examples available that I can use to effectively generate what should be completely valid tests for you to use. While you can contact me on github through these issues, let me know if you want to try discussing them in a different way that might be easier, and let me know if you are missing test files for anything. Chances are I can make them.

**Comment by seamustuohy** [comment]
@TheElementalOfDestruction I believe I've addressed this issue in the [major_structural_changes](https://github.com/seamustuohy/RTFDE/tree/major_structural_changes) branch.  If you have a test file you should be able to verify by using the ["private data test folder" functionality described here](https://github.com/seamustuohy/RTFDE/tree/major_structural_changes/tests/test_data/personal_rtf_test_files).

**Comment by seamustuohy** [comment]
Addressed in the 0.1.0 version which is in the final notification and testing stage on [dev](https://github.com/seamustuohy/RTFDE/commit/1b29704a47e6c3a118ebb3ed2ac96e640312f8fb)
